### PR TITLE
zrythm: 1.0.0-alpha.28.1.3 -> 1.0.0-beta.4.5.62; carla: 2.5.1 -> 2.5.3

### PR DESCRIPTION
**Pull request by yuuyins**
###### Description of changes

zrythm: 1.0.0-alpha.28.1.3 
carla: 2.5.1 -> 2.5.3
zix: init at unstable-2023-02-13

supersedes
- https://github.com/NixOS/nixpkgs/pull/190730
- https://github.com/NixOS/nixpkgs/pull/175505
- https://github.com/NixOS/nixpkgs/pull/196227

tested and works

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cole-h** [comment]
@ofborg eval

**Review by magnetophon — APPROVED** [review]

File: pkgs/applications/audio/carla/default.nix [hunk 1]
```diff
@@ -13,13 +13,13 @@ assert withGtk3 -> gtk3 != null;
 
 stdenv.mkDerivation rec {
   pname = "carla";
-  version = "2.5.1";
+  version = "unstable-2023-02-12";
```
Comment (magnetophon) [inline pkgs/applications/audio/carla/default.nix:16]: Did https://github.com/falkTX/Carla/releases/tag/v2.5.3 not work?

**Comment by magnetophon** [comment]
Maybe it would be best to make Carla unstable a separate pkg.

**Review by magnetophon — APPROVED** [review]
Looks great, thanks!

**Review by fogti — APPROVED** [review]

File: pkgs/development/libraries/audio/zix/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  version = "unstable-2023-02-13";
+
+  src = fetchFromGitLab {
+    owner = "drobilla";
+    repo = pname;
+    rev = "262d4a1522c38be0588746e874159da5c7bb457d";
+    hash = "sha256-3vuefgnirM4ksK3j9sjBHgOmx0JpL+6tCPb69/7jI00=";
+  };
+in stdenv.mkDerivation {
+  inherit pname version src;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/audio/zix/default.nix:20]:
```suggestion
stdenv.mkDerivation rec {
  pname = "zix";
  version = "unstable-2023-02-13";

  src = fetchFromGitLab {
    owner = "drobilla";
    repo = pname;
    rev = "262d4a1522c38be0588746e874159da5c7bb457d";
    hash = "sha256-3vuefgnirM4ksK3j9sjBHgOmx0JpL+6tCPb69/7jI00=";
  };
```
we can make this a bit more straight forward

File: pkgs/applications/audio/zrythm/default.nix [hunk 3]
```diff
@@ -161,13 +192,17 @@ stdenv.mkDerivation rec {
...
 
   meta = with lib; {
     homepage = "https://www.zrythm.org";
-    description = "Highly automated and intuitive digital audio workstation";
-    maintainers = with maintainers; [ tshaynik magnetophon ];
+    description = "Automated and intuitive digital audio workstation";
+    maintainers = with maintainers; [
+      tshaynik
+      magnetophon
+    ];
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zrythm/default.nix:205]:
```suggestion
    maintainers = with maintainers; [ tshaynik magnetophon ];
```
nit
we don't need to change this. Doing it like this is fine.

File: pkgs/applications/audio/zrythm/default.nix [hunk 4]
```diff
@@ -161,13 +192,17 @@ stdenv.mkDerivation rec {
     gappsWrapperArgs+=(
       --prefix GSETTINGS_SCHEMA_DIR : "$out/share/gsettings-schemas/${pname}-${version}/glib-2.0/schemas/"
       --prefix XDG_DATA_DIRS : "$XDG_ICON_DIRS"
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zrythm/default.nix:194]:
```suggestion
      --prefix XDG_DATA_DIRS : "$XDG_ICON_DIRS:${breeze-icons}/share"
```
we can combine this to generate less code

File: pkgs/applications/audio/zrythm/default.nix [hunk 5]
```diff
@@ -122,36 +139,50 @@ stdenv.mkDerivation rec {
     rtaudio
     rtmidi
     rubberband
+    sassc
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/zrythm/default.nix:142]: Isn't sassc a css compiler that is run on the host?

File: pkgs/applications/audio/carla/default.nix [hunk 6]
```diff
@@ -1,91 +1,13 @@
...
-      It uses JACK as the default and preferred audio driver but also
-      supports native drivers like ALSA, DirectSound or CoreAudio.
-    '';
-    license = licenses.gpl2Plus;
-    maintainers = [ maintainers.minijackson ];
-    platforms = platforms.linux;
+  carla-unstable = libsForQt5.callPackage ./common.nix {
+    version = "unstable-2023-02-12";
+    hash = "sha256-OegwW8sfUV4IjBFrq38tpKbOMzZT0cvhL/u3Gr1m938=";
+    rev = "4c29367b6e6f82e75746b9ee3a01f9d205c0beae";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/carla/default.nix:11]: Do we only need this for one package? Then we should not expose it on top-level and rather just do the overwrite for the one package

File: pkgs/applications/audio/carla/common.nix [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  ];
+
+  pythonPath = with python3Packages; [
+    rdflib pyliblo
+  ] ++ lib.optional withFrontend pyqt5;
+
+  buildInputs = [
+    file liblo alsa-lib fluidsynth jack2 libpulseaudio libsndfile
+  ] ++ lib.optional withQt qtbase
+    ++ lib.optional withGtk2 gtk2
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/carla/common.nix:55]: We should think about removing gtk2

File: pkgs/applications/audio/carla/common.nix [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+, version
+, rev
+, hash
+}:
+
+assert withFrontend -> python3Packages ? pyqt5;
+assert withQt -> qtbase != null;
+assert withQt -> wrapQtAppsHook != null;
+assert withGtk2 -> gtk2 != null;
+assert withGtk3 -> gtk3 != null;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/carla/common.nix:31]:
```suggestion
```
that will never be the case in nixpkgs

File: pkgs/applications/audio/carla/common.nix [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+, withGtk2 ? true
+, withGtk3 ? true
+, withQt ? true
+, wrapQtAppsHook ? null
+, version
+, rev
+, hash
+}:
+
+assert withFrontend -> python3Packages ? pyqt5;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/carla/common.nix:27]:
```suggestion
```
same here

File: pkgs/applications/audio/zrythm/default.nix (hunk 5 above)
Comment (zseri) [reply pkgs/applications/audio/zrythm/default.nix:155]: no I think this is about why it is in `buildInputs` and not in `nativeBuildInputs`

File: pkgs/development/libraries/audio/zix/default.nix (hunk 2 above)
Comment (Lord-Valen) [reply pkgs/development/libraries/audio/zix/default.nix:20]: If you don't want to use `rec` you could also do
```nix
let
  name = "zix";
in stdenv.mkDerivation rec {
  pname = name;
  version = "unstable-2023-02-13";
  src = fetchFromGitLab {
    owner = "drobilla";
    repo = name;
    rev = "262d4a1522c38be0588746e874159da5c7bb457d";
    hash = "sha256-3vuefgnirM4ksK3j9sjBHgOmx0JpL+6tCPb69/7jI00=";
  };
```

Comment (Lord-Valen) [reply pkgs/development/libraries/audio/zix/default.nix:20]: > i suppose it was not your intention to have left rec in in stdenv.mkDerivation rec { in your suggestion?

Oh I didn't notice that I had left `rec` in.	Oops!

> i can't see how the suggestion would improve over inherit (i mean the pname=name bits)

In my opinion, minimizing the contents of a `let` block is typically more readable. It's more immediately apparent what is being given to `mkDerivation`.

> i'm using src.rev on meta.changelog as well for this derivation, and i could always have to use version as well...

If you want to also use rev, you could hoist that up to the `let` block as well.

> i know this rec / no rec may be a minor thing, but I really want to avoid anything that might cause those nix issues.

You're using the `let` block like a `rec` here *anyway* (you're using it for self-reference). The end result was that you added extra boilerplate/clutter (the `inherit`s) to no added benefit. That's why Sandro recommended those changes. Arguably, this use of a `let` block is even more of an anti-pattern than `rec` since it's really just `rec` with extra steps. If you look closely, what nix.dev was suggesting as an alternative to `rec` was exactly the sort of thing I suggested: extract the commonalities to new variables, and put those variables in scope with a `let` block.

All that being said, there isn't really a risk of inifinite recursion using `rec` this way. Either go for the `rec` as Sandro suggested, or use my suggested `let` block. Both are acceptable imo.

**Review by Lord-Valen — APPROVED** [review]
Looks good!

**Review by Lord-Valen — APPROVED** [review]

File: pkgs/applications/audio/carla/default.nix [hunk 10]
```diff
@@ -1,25 +1,38 @@
...
-assert withGtk3 -> gtk3 != null;
 
 stdenv.mkDerivation rec {
   pname = "carla";
-  version = "2.5.1";
+  version = "2.5.3";
 
   src = fetchFromGitHub {
-    owner = "falkTX";
+    owner = "falktx";
```
Comment (zseri) [inline pkgs/applications/audio/carla/default.nix:32]: why was this changed?

File: pkgs/applications/audio/zrythm/default.nix [hunk 11]
```diff
@@ -46,31 +55,48 @@
...
+let
+  # As of zrythm-1.0.0-beta.4.5.62, Zrythm needs clap
+  # https://github.com/falktx/carla/tree/main/source/includes/clap, which is
+  # only available on Carla unstable as of 2023-02-24.
+  carla-unstable = carla.overrideAttrs (oldAttrs: rec {
+    pname = "carla";
+    version = "unstable-2023-02-24";
+
+    src = fetchFromGitHub {
+      owner = "falktx";
```
Comment (zseri) [inline pkgs/applications/audio/zrythm/default.nix:81]:
```suggestion
      owner = "falkTX";
```

**Review by magnetophon — APPROVED** [review]

**Review by Lord-Valen — APPROVED** [review]

**Review by minijackson — APPROVED** [review]
Works great for carla, thanks!

**Review by fogti — APPROVED** [review]

File: pkgs/applications/audio/carla/default.nix [hunk 12]
```diff
@@ -1,25 +1,38 @@
...
+, fluidsynth
+, jack2
+, liblo
+, libpulseaudio
+, libsndfile
+, pkg-config
+, python3Packages
+, which
+, gtk2 ? null
+, gtk3 ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/carla/default.nix:15]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please move the gitignore changes to another PR unless they are directly related to the updated packages but then we might as well put a gitignore file into that specific directory.
Changes to that file are relevant for a lot of people and I cannot just merge them in a package bump.

**Review by tshaynik — APPROVED** [review]

File: pkgs/applications/audio/carla/common.nix (hunk 7 above)
Comment (minijackson) [reply pkgs/applications/audio/carla/common.nix]: IIRC, gtk2 is needed for plugins whose interface is implemented using gtk2. I can't think of an example plugin right now, but removing it might break the setup of some people using gtk2-based audio plugins

**Review by Lord-Valen — APPROVED** [review]

**Review by fogti — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]
